docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,301 @@
+
+Adding `pruneBundle` should ensure a mechanism for deleting `docker.io/<registry-repository>/bundle:latest`.
+
+```yaml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-kaniko-build
+spec:
+  source:
+    container:
+      image: docker.io/<registry-repository>/bundle:latest
+      pruneBundle: true # Disable by default
+  strategy:
+    name: kaniko
+    kind: ClusterBuildStrategy
+  output: ...
+```
+
+The above Build definition should ensure that a step in our [TaskSpec](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/reconciler/buildrun/resources/taskrun.go#L59) definition is appended, which will ensure a pruning of the image in the related container registry.
+
+_Note_: Supporting multiple registry providers might be too complicated and different alternatives should be considering for pruning.
+
+#### As a Shipwright/CLI contributor I want to provide a mechanism to support local source code
+
+This is related to the CLI workflow, that should support the bundle approach and maintain the bundle base image. The implementation should be generic enough for other workflows, like OpenShift Build, IBM Cloud Code Engine, to consume it.
+
+### Implementation Details/Notes/Constraints [optional]
+
+There is a prototype for educational purposes that can be tested if desire, see the related [documentation](https://github.com/qu1queee/cli/blob/qu1queee/crud_cmd/README.md). This uses the following two branches:
+
+- Changes in Shipwright/Build, see [branch](https://github.com/qu1queee/build/tree/qu1queee/local_source)
+- Changes in Shipwright/CLI, see [branch](https://github.com/qu1queee/cli/tree/qu1queee/crud_cmd)
+
+### Risks and Mitigations
+
+- This requires coordination between Build and CLI. A well defined list of issues for both backlogs should be created.
+
+- Using a container registry to push/pull the source code might bring concerns around performance. This EP only covers functional requirements. Non-functional requirements like performance or scalability should be discuss during the proposed implementation, and via further development cycles.
+
+- The approach of bundles should not lock-in the CLI with a single approach. It should be flexible enough, to extend to new approaches in the future.
+
+## Design Details
+
+### Test Plan
+
+- Requires development of the self-extractable base image in Shipwright/CLI.
+- Requires development of the bundles `pkg` in Shipwright/CLI.
+- Requires development of the API extension and the runtime changes in Shipwright/Build.
+- This requires new unit, integration and a new e2e tests with local source code in both CLI and Build.
+
+### Graduation Criteria
+
+Should be part of any release before our v1.0.0
+
+### Upgrade / Downgrade Strategy
+
+Not apply, this should not break anything, is rather an extension of the API.
+
+### Version Skew Strategy
+
+N/A
+
+## Implementation History
+
+N/A
+
+## Drawbacks
+
+None
```

Comment:
IMO the most significant drawback is that this approach relies on a container registry being present, and that this container registry is safe for users to push source code to. Not all clusters have a container registry inside it, not all developers have direct access to a container registry. We should make this clear here.

As you noted in the Alternatives section, this EP does not preclude other approaches being added in the future.

Reply:
I will add that soon, thanks.